src/Dialect/ONNX/DialectBuilder.cpp

```diff
@@ -50,29 +50,44 @@ Value OnnxBuilder::matmul(Type Y, Value A, Value B, bool useGemm) const {
                     B.getType().isa<ShapedType>() &&
                     B.getType().cast<ShapedType>().hasRank() &&
                     (B.getType().cast<ShapedType>().getRank() == 2);
+  auto aValue = totensor(A);
+  auto bValue = totensor(B);
   if (canUseGemm)
-    return b.create<ONNXGemmOp>(loc, Y, A, B, b.createOrFold<ONNXNoneOp>(loc),
+    return b.create<ONNXGemmOp>(loc, Y, aValue, bValue,
+        b.createOrFold<ONNXNoneOp>(loc),
         /*alpha=*/b.getF32FloatAttr(1.0), /*beta=*/b.getF32FloatAttr(1.0),
         /*transA=*/
         IntegerAttr::get(b.getIntegerType(64, /*isSigned=*/true),
             APInt(64, 0, /*isSigned=*/true)),
         /*transB=*/
         IntegerAttr::get(b.getIntegerType(64, /*isSigned=*/true),
             APInt(64, 0, /*isSigned=*/true)));
-  return b.create<ONNXMatMulOp>(loc, Y, A, B);
+  return b.create<ONNXMatMulOp>(loc, Y, aValue, bValue);
 }
 
 Value OnnxBuilder::reshape(Type outputType, Value input, Value shape) const {
-  return b.create<ONNXReshapeOp>(loc, outputType, input, shape);
+  return b.create<ONNXReshapeOp>(
+      loc, outputType, totensor(input), totensor(shape));
 }
 
 Value OnnxBuilder::transpose(
     Type outputType, Value input, ArrayAttr perm) const {
-  return b.create<ONNXTransposeOp>(loc, outputType, input, perm);
+  return b.create<ONNXTransposeOp>(loc, outputType, totensor(input), perm);
 }
 
 Value OnnxBuilder::constant(Attribute denseAttr) const {
   return b.create<ONNXConstantOp>(loc, Attribute(), denseAttr);
 }
 
+Value OnnxBuilder::totensor(Value input) const {
+  if (input.getType().isa<TensorType>())
+    return input;
+  assert(input.getType().isa<MemRefType>() &&
+         "expect RankedMemref type when not a TensorType");
+  auto aTy = input.getType().cast<ShapedType>();
+  auto aTensorTy = RankedTensorType::get(aTy.getShape(), aTy.getElementType());
+  return b.create<UnrealizedConversionCastOp>(loc, aTensorTy, input)
+      .getResult(0);
+}
```

Comment:
It looks non-trivial to convert MemRefType to TensorType when MemRefType has non-trivial affine_map. I think we will use MemRefType with affine_map more often in near future. So we should make sure we have no problem with that.

Reply:
Tensor type should be inferred from the type of the original inputs (of Tensor type) of ONNX Op, not the transformed ones (Operands in the conversion). Inference tensor type from memref type will be avoided.

Reply:
> Tensor type should be inferred from the type of the original inputs (of Tensor type) of ONNX Op, not the transformed ones (Operands in the conversion)

I see. Thanks!